# Exploratory Data Analysis
## Data Overview
**Goal**: Understand the structure of the dataset, including the number of rows, columns, and data types.

**Key Steps**: Load the data and print a concise summary of the dataset using .info() and .head().

In [13]:
#Import necessary libraries
import pandas as pd
import logging
import os, sys
# Add the 'scripts' directory to the Python path for module imports
sys.path.append(os.path.abspath(os.path.join('..', 'scripts')))
# Import load_data module
from load_data import load_data # type: ignore

# Set max rows and columns to display
pd.set_option('display.max_columns', 200)
pd.set_option('display.max_rows', 200)

# Configure logging
from custom_logger import setup_logger # type: ignore
logger = setup_logger()

logger.info("Imported necessary libraries.")

2025-01-23 22:46:59,741 - INFO - Imported necessary libraries.


Load the dataset

In [14]:
logger.info("Data loading initiated.")
df = load_data('../data/data.csv')  # Assume load_data() is your function
logger.info("Data loaded successfully.")

2025-01-23 22:46:59,755 - INFO - Data loading initiated.
2025-01-23 22:47:00,006 - INFO - Data loaded successfully.


Data successfully loaded from ../data/data.csv
Dataset contains 95662 rows and 15 columns.



### 1. Data overview:
Provide an overview of the dataset including shape, data types, missing values, and first few rows.

In [15]:
# Import the class CreditRiskEDA
from credit_eda_analysis import CreditRiskEDA
# Initialize the class
cr_eda = CreditRiskEDA(df)
# Logging activity
logger.info("Data overview initiated.")

# Overview of the dataset
if not df.empty:
    cr_eda.data_overview()

logger.info("Data overview successfully completed.")

2025-01-23 22:47:00,014 - INFO - Data overview initiated.


Data Overview:
Number of rows: 95662
Number of columns: 15

Column Data Types:
BatchId                  object
AccountId                object
SubscriptionId           object
CustomerId               object
CurrencyCode             object
CountryCode               int64
ProviderId               object
ProductId                object
ProductCategory          object
ChannelId                object
Amount                  float64
Value                     int64
TransactionStartTime     object
PricingStrategy           int64
FraudResult               int64
dtype: object

First Five Rows:


BatchId       AccountId       SubscriptionId  \
TransactionId                                                              
TransactionId_76871   BatchId_36123  AccountId_3957   SubscriptionId_887   
TransactionId_73770   BatchId_15642  AccountId_4841  SubscriptionId_3829   
TransactionId_26203   BatchId_53941  AccountId_4229   SubscriptionId_222   
TransactionId_380    BatchId_102363   AccountId_648  SubscriptionId_2185   
TransactionId_28195   BatchId_38780  AccountId_4841  SubscriptionId_3829   

                          CustomerId CurrencyCode  CountryCode    ProviderId  \
TransactionId                                                                  
TransactionId_76871  CustomerId_4406          UGX          256  ProviderId_6   
TransactionId_73770  CustomerId_4406          UGX          256  ProviderId_4   
TransactionId_26203  CustomerId_4683          UGX          256  ProviderId_6   
TransactionId_380     CustomerId_988          UGX          256  ProviderId_1   
TransactionId_28195   CustomerId_988          UGX          256  ProviderId_4   

                        ProductId     ProductCategory    ChannelId   Amount  \
TransactionId                                                                 
TransactionId_76871  ProductId_10             airtime  ChannelId_3   1000.0   
TransactionId_73770   ProductId_6  financial_services  ChannelId_2    -20.0   
TransactionId_26203   ProductId_1             airtime  ChannelId_3    500.0   
TransactionId_380    ProductId_21        utility_bill  ChannelId_3  20000.0   
TransactionId_28195   ProductId_6  financial_services  ChannelId_2   -644.0   

                     Value  TransactionStartTime  PricingStrategy  FraudResult  
TransactionId                                                                   
TransactionId_76871   1000  2018-11-15T02:18:49Z                2            0  
TransactionId_73770     20  2018-11-15T02:19:08Z                2            0  
TransactionId_26203    500  2018-11-15T02:44:21Z                2            0  
TransactionId_380    21800  2018-11-15T03:32:55Z                2            0  
TransactionId_28195    644  2018-11-15T03:34:21Z                2            0

2025-01-23 22:47:00,048 - INFO - Data overview successfully completed.



Missing Values Overview:
BatchId                 0
AccountId               0
SubscriptionId          0
CustomerId              0
CurrencyCode            0
CountryCode             0
ProviderId              0
ProductId               0
ProductCategory         0
ChannelId               0
Amount                  0
Value                   0
TransactionStartTime    0
PricingStrategy         0
FraudResult             0
dtype: int64


### Data Overview Summary

+ The dataset contains 95,662 rows and 15 columns, providing a substantial amount of data for analysis.

+ The data types are generally correct, except for the TransactionStartTime column, which is currently stored as an object. This column should be converted to datetime format for accurate time-based analysis.

+ No missing values were detected in any of the columns, ensuring the dataset is complete and ready for further analysis without the need for imputation.

In [16]:
# Convert the TransactionStartTime to appropriate datetime format
df['TransactionStartTime'] = pd.to_datetime(df['TransactionStartTime'])

In [17]:
df.to_csv("../data/output.csv", index=False)

OSError: Cannot save file into a non-existent directory: 'data'

### 2. Summary Statistics

Understand the central tendency, dispersion, and shape of the dataset’s distribution.

In [ ]:
# Log the start of the summary statistics process
logger.info("Generating summary statistics for numeric columns...")   

# Statistical summary
summary_stats = cr_eda.summary_statistics()
display(summary_stats.T)
# Log completion
logger.info("Summary statistics generation completed.")

### Observations:

+ The CountryCode is constant at 256, indicating no variability in this column.

+ The Amount and Value features show a high degree of skewness (both over 51), indicating potential outliers or heavy-tailed distributions.

+ PricingStrategy is fairly stable with most values centered around 2.

+ FraudResult has a very low mean, indicating a low occurrence of fraudulent transactions, with a highly positive skewness suggesting a majority of zeros.

### Distribution of Numerical Features

+ Visualize the distribution of numerical features to identify patterns, skewness, and potential outliers.

In [ ]:
# Import the class CreditRiskEDA
from credit_eda_visualize import CreditRiskEDAVisualize
# Initialize the class
cr_eda_visual = CreditRiskEDAVisualize(df)
# Logger activity
logger.info("Plot Numberical Distribution...")
# List of numeric columns
numeric_cols = df.select_dtypes(include='number').columns

# Plot distibution
cr_eda_visual.plot_numerical_distribution(numeric_cols)
logger.info("The distribution plot successfully completed.")

### Key insights:

+ Most of the numerical features exhibit right-skewness, indicating that there are a few extreme values that pull the mean to the right.

+ The distributions of CountryCode, Amount, and Value have clear peaks, suggesting common values or ranges.

+ The distribution of PricingStrategy has a clear peak at 2, indicating a preferred pricing strategy.

+ The distribution of FraudResult shows a majority of non-fraudulent transactions.

### Skewness:

+ Visualize the skewness of numerical features

In [ ]:
# Logger activity
logger.info("Visualizing Skewness of numercial features...")
if not df.empty:
    cr_eda_visual.plot_skewness()
logger.info("Visualization successfully completed.")

### Key Insights:

+ CountryCode: This feature remains constant (256) across all entries, indicating the data pertains to a single geographical region. Since there is no variability in this feature, it will not contribute to model performance and can be considered for removal in subsequent analysis.

+ Amount: The distribution of the transaction amount is highly right-skewed with extreme outliers, as indicated by a skewness value of over 51. This suggests that a few transactions have significantly higher amounts than the rest. Such skewness can distort the model's learning process, so normalization or transformation (e.g., log transformation) may be necessary to handle these extreme values.

+ Value: The absolute value of transactions shows a similar pattern to Amount, with a high degree of skewness. The presence of large outliers implies that these transactions have a disproportionate effect on the overall data. Addressing this skewness through appropriate scaling will help balance the data for better model performance.

+ PricingStrategy: The data shows a preference for a specific pricing strategy (category 2), which dominates the distribution. This could indicate a business preference or customer inclination toward that pricing strategy. However, there is less skewness here, suggesting that the distribution is more balanced compared to other features.

+ FraudResult: Fraudulent transactions are extremely rare, as evidenced by the mean close to zero and the right-skewness of over 22. This class imbalance may lead to model bias toward predicting non-fraudulent transactions. Addressing this imbalance through techniques like oversampling the fraud cases or employing specialized algorithms may be required to ensure the model effectively captures fraudulent behavior.

#### Distribution of Categorical Features

+ Analyzing the distribution of categorical features provides insights into the frequency and variability of categories.

In [ ]:
logger.info('Visualizing the distribution of useful categorical features...')
if not df.empty:
    cr_eda_visual.plot_categorical_distribution()
logger.info('Distribution plot successfully completed.')

### Key Insights from Categorical Features:

**CurrencyCode:**

The dataset is dominated by a single currency, UGX (Ugandan Shilling), which is the currency for all transactions. This uniformity indicates that the platform primarily handles transactions within Uganda. Since CurrencyCode is uniform across the entire dataset (with only UGX as the value), it does not provide any variability or discriminative power for future model training. As a result, it would not contribute meaningfully to the predictive model and can be excluded from the feature set during model development.

**ProviderId:**

Two providers, ProviderId_6 and ProviderId_4, handle the majority of transactions, with counts of 34,186 and 38,189 respectively. This suggests a concentrated reliance on these providers for most services. Other providers have much smaller transaction volumes, with ProviderId_1 and ProviderId_5 contributing some, while ProviderId_2 has an almost negligible number of transactions.

**ProductCategory:**

Two product categories, airtime (45,027) and financial_services (45,405), dominate the dataset. These are clearly the most popular categories, suggesting a focus on telecom and financial services transactions. Other categories like utility_bill, data_bundles, and tv have significantly fewer transactions, indicating less customer demand or offering.

**ChannelId:**

The majority of transactions were made through ChannelId_3 (56,935) and ChannelId_2 (37,141), indicating these are the most popular channels. Other channels, such as ChannelId_1 and ChannelId_5, are used much less frequently.

### Correlation Analysis

+ Understanding the relationship between numerical features.

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt
# Select numerical columns for correlation analysis
numerical_cols = df.select_dtypes(include=['float64', 'int64']).columns

# Calculate the correlation matrix
correlation_matrix = df[numerical_cols].corr()

# Set up the matplotlib figure
plt.figure(figsize=(12, 8))

# Draw the heatmap with the mask and correct aspect ratio
sns.heatmap(correlation_matrix, annot=True, fmt=".2f", cmap='coolwarm', square=True, cbar_kws={"shrink": .8}, linewidths=0.5)

# Title and labels
plt.title('Correlation Matrix of Numerical Features', fontsize=16)
plt.xticks(rotation=45)
plt.yticks(rotation=45)
plt.tight_layout()

+ Amount and Value has strong positive correlation since they provide the same information. Value has no significance for future modeling

+ Amount and FraudResults has moderate correlation

+ PricingStrategy and Other features has no correlation

### Outlier Detection

+ Use box plots to identify outliers.

In [ ]:
cr_eda_visual.detect_outliers(['Amount', 'Value', 'FraudResult'])

The box plots show the distribution of numerical features along with the outliers:

+ Amount and Value: Both show significant numbers of outliers, particularly with high transaction values. These outliers could represent extreme transaction cases or potential fraudulent activity, requiring further investigation or possible scaling during preprocessing.

+ FraudResult: Outliers here are limited, likely because this is a binary variable, but it is crucial to note the imbalanced nature of fraud occurrences.

**Recommendations:**

+ Handling Outliers: Consider applying robust scaling methods or transformations like logarithmic scaling to normalize the distribution of Amount and Value.

### Boxplot of Transaction Amounts for Fraud and Non-Fraud Transactions

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

# Creating a plot to visualize the relationship between Amount and FraudResult
plt.figure(figsize=(10, 4))

# Plotting boxplot to compare Amount for FraudResult = 1 (fraudulent) and FraudResult = 0 (non-fraudulent)
sns.boxplot(x='FraudResult', y='Amount', hue='FraudResult', legend=False,data=df, palette="Set2")

plt.title('Boxplot of Transaction Amounts for Fraud and Non-Fraud Transactions')
plt.xlabel('Fraud Result (0 = Non-Fraud, 1 = Fraud)')
plt.ylabel('Transaction Amount')
plt.yscale('log')  # Using log scale to handle the large range of amounts

plt.show()

### Cross-tabulate Fraud with Amount Ranges

+ Divide the Amount variable into ranges (bins) and create a cross-tabulation with FraudResult. This will show how often high-value transactions (those in the top bins) are labeled as frauds.

In [ ]:
# Define bins for Amount ranges
bins = [0, 1000, 5000, 10000, 50000, 100000, df['Amount'].max()]
labels = ['<1K', '1K-5K', '5K-10K', '10K-50K', '50K-100K', '>100K']
df_new = df.copy()
# Create a new column with binned Amount
df_new['AmountRange'] = pd.cut(df_new['Amount'], bins=bins, labels=labels, include_lowest=True)

# Create a cross-tabulation
cross_tab = pd.crosstab(df_new['AmountRange'], df_new['FraudResult'], margins=True, margins_name="Total")
print(cross_tab)

# Optional: Visualize the cross-tabulation with a heatmap
plt.figure(figsize=(8, 4))
sns.heatmap(cross_tab, annot=True, cmap='coolwarm', fmt='g')
plt.title('Cross Tabulation of Amount Ranges and Fraud Result')
plt.show()

### Key insights from the cross-tabulation

1. Low-value transactions (<1K): Nearly all transactions in this range are non-fraudulent, suggesting low risk.

2. Medium-value transactions (1K-50K): Mid-range transactions also show low fraud rates, with just 3 frauds in the 10K-50K range.

3. High-value transactions (50K-100K): Fraud increases to about 1.39% in this range, indicating higher risk.

4. Very high-value transactions (>100K): Around 26.7% of these transactions are fraudulent, showing a strong link between large amounts and fraud.

+ Outliers should be retained, as they highlight a significant fraud risk in high-value transactions.
+ The model could be refined to place more emphasis on high-value transactions, where fraud is more prevalent.

### Save cleaned data for feature engineering analysis

In [ ]:
#save to data/output.csv.csv
df.to_csv("data/output.csv",index = False)